File: Visitor/Visitor.php
<?php

/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 19:34
 */
interface ComputerPart
{
    public function accept($computerPartVisitor);
}

class Keyboard implements ComputerPart
{
    public function accept($computerPartVisitor)
    {
        // TODO: Implement accept() method.
        $computerPartVisitor->visit($this);
    }
}


class Monitor implements ComputerPart
{
    public function accept($computerPartVisitor)
    {
        // TODO: Implement accept() method.
        $computerPartVisitor->visit($this);
    }
}

class Mouse implements ComputerPart
{
    public function accept($computerPartVisitor)
    {
        // TODO: Implement accept() method.
        $computerPartVisitor->visit($this);
    }
}


class Computer implements ComputerPart
{
    private $parts;

    public function __construct()
    {
        $this->parts=array(new Mouse(),new Keyboard(),new Monitor());
    }

    public function accept($computerPartVisitor)
    {
        // TODO: Implement accept() method.
        foreach($this->parts as $part){
            $part->accept($computerPartVisitor);
        }

        $computerPartVisitor->visit($this);
    }
}

interface computerPartVisitor{
    public function visit($part);
}

class ComputerPartDisplayVisitor  implements computerPartVisitor{
    public function visit($part)
    {
        // TODO: Implement visit() method.
        if(!is_object($part)) var_dump("not found part");
        $partName=get_class($part);
        switch ($partName){
            case "Computer":
                var_dump("Displaying Computer.");
                break;
            case "Mouse":
                var_dump("Displaying Mouse.");
                break;
            case "Keyboard":
                var_dump("Displaying Keyboard.");
                break;
            case "Monitor":
                var_dump("Displaying Monitor.");
                break;
        }
    }
}

//demo
$computer=new Computer();
$computer->accept(new ComputerPartDisplayVisitor());

File: Visitor/Visitor.go
package main

import (
	"fmt"
)

type ComputerPartVister struct {

}

func (c *ComputerPartVister)visit(part ComputerPart) {
	switch part.(type){
	case *Mouse:
		fmt.Println("Displaying Mouse.")
	case *Monitor:
		fmt.Println("Displaying Monitor.")
	case *Keyboard:
		fmt.Println("Displaying Keyboard.")
	case *Computer:
		fmt.Println("Displaying Computer.")
	}
}

type ComputerPart interface {
	accept(visitor *ComputerPartVister)
}

type Mouse struct {

}

func (m *Mouse)accept(visitor *ComputerPartVister) {
	visitor.visit(m)
}

type Keyboard struct {

}

func (m *Keyboard)accept(visitor *ComputerPartVister) {
	visitor.visit(m)
}

type Monitor struct {

}

func (m *Monitor)accept(visitor *ComputerPartVister) {
	visitor.visit(m)
}

type Computer struct {
	parts []ComputerPart
}

func (c *Computer)accept(visitor *ComputerPartVister) {
	for _, part := range c.parts {
		part.accept(visitor)
	}

	visitor.visit(c)
}

func main() {
	parts := []ComputerPart{new(Monitor), new(Mouse), new(Keyboard)}
	computer := &Computer{parts}
	computer.accept(new(ComputerPartVister))
}


